### PR TITLE
Cached path command

### PR DESCRIPTION
**Pull request by epwalsh**
Usage examples:

```bash
> allennlp cached-path --help
> allennlp cached-path https://storage.googleapis.com/allennlp-public-models/sst-roberta-large-2020.06.08.tar.gz
> allennlp cached-path --extract-archive https://storage.googleapis.com/allennlp-public-models/sst-roberta-large-2020.06.08.tar.gz
> allennlp cached-path --inspect
> allennlp cached-path --inspect 'https://storage.google*'
> allennlp cached-path --remove 'https://storage.google*'
```

[asciicast](https://asciinema.org/a/Ab02Ud0rNAB1e9iutS3DxCMad)

### CODE REVIEW

File: allennlp/common/file_utils.py [hunk 1]
```diff
@@ -509,3 +599,162 @@ def text_lines_from_file(filename: Union[str, Path], strip_lines: bool = True) -
...
+    for resource, metas in sorted(
+        cache_entries.items(),
+        # Sort by creation time, latest first.
+        key=lambda x: max(
+            0 if not x[1][0] else x[1][0][0].creation_time,
+            0 if not x[1][1] else x[1][1][0].creation_time,
+        ),
+        reverse=True,
+    ):
+        print("\n➥", resource)
```
Comment (dirkgr) [inline allennlp/common/file_utils.py:743]: Special character in a string literal?

**Review by dirkgr — APPROVED** [review]
This is very close to being its own module. Maybe even its own library. Though I'd almost say "Contributions Welcome" for making it its own library, because that's going to be a lot of work without a ton of immediate benefit to AllenNLP itself.

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added `FBetaMultiLabelMeasure`, a multi-label Fbeta metric. This is a subclass of the existing `FBetaMeasure`.
 - Added ability to pass additional key word arguments to `cached_transformers.get()`, which will be passed on to `AutoModel.from_pretrained()`.
 - Added an `overrides` argument to `Predictor.from_path()`.
+- Added a `cached-path` command.
+- Added a function `inspect_cache` to `common.file_utils` that prints useful information about the cache. This can also 
+  be used from the `cached-path` command with `allennlp cached-path --inspect`.
+- Added a function `remove_cache_entries` to `common.file_utils` that removes any cache entries matching the given
+  glob patterns. This can used from the `cached-path` command with `allennlp cached-path --remove some-files-*`.
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply CHANGELOG.md:26]: That's already implemented in the Rust version 😉  https://github.com/epwalsh/rust-cached-path/blob/master/src/cache.rs#L97

File: allennlp/common/file_utils.py [hunk 3]
```diff
@@ -509,3 +599,162 @@ def text_lines_from_file(filename: Union[str, Path], strip_lines: bool = True) -
...
+    The values in the returned mapping are tuples because we seperate meta entries that
+    correspond to extraction directories vs regular cache entries.
+    """
+    cache_dir = os.path.expanduser(cache_dir or CACHE_DIRECTORY)
+
+    total_size: int = 0
+    cache_entries: Dict[str, Tuple[List[_Meta], List[_Meta]]] = defaultdict(lambda: ([], []))
+    for meta_path in glob.glob(str(cache_dir) + "/*.json"):
+        meta = _Meta.from_path(meta_path)
+        if patterns and not any(fnmatch(meta.resource, p) for p in patterns):
```
Comment (dirkgr) [inline allennlp/common/file_utils.py:679]: aDwMMBN_700b

File: allennlp/common/file_utils.py [hunk 4]
```diff
@@ -509,3 +599,162 @@ def text_lines_from_file(filename: Union[str, Path], strip_lines: bool = True) -
...
+    Returns the total size of the matching entries and mapping or resource name to meta data.
+
+    The values in the returned mapping are tuples because we seperate meta entries that
+    correspond to extraction directories vs regular cache entries.
+    """
+    cache_dir = os.path.expanduser(cache_dir or CACHE_DIRECTORY)
+
+    total_size: int = 0
+    cache_entries: Dict[str, Tuple[List[_Meta], List[_Meta]]] = defaultdict(lambda: ([], []))
+    for meta_path in glob.glob(str(cache_dir) + "/*.json"):
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply allennlp/common/file_utils.py:677]: The extension gets removed when we cached a file, so only meta files should have `.json` extensions.

File: allennlp/common/file_utils.py [hunk 5]
```diff
@@ -509,3 +599,162 @@ def text_lines_from_file(filename: Union[str, Path], strip_lines: bool = True) -
...
+    patterns: List[str] = None,
+    cache_dir: Union[str, Path] = None,
+) -> Tuple[int, Dict[str, Tuple[List[_Meta], List[_Meta]]]]:
+    """
+    Find all cache entries, filtering ones that don't match any of the glob patterns given.
+
+    Returns the total size of the matching entries and mapping or resource name to meta data.
+
+    The values in the returned mapping are tuples because we seperate meta entries that
+    correspond to extraction directories vs regular cache entries.
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply allennlp/common/file_utils.py:671]: Yes that would probably be cleaner.

File: allennlp/common/file_utils.py [hunk 6]
```diff
@@ -509,3 +599,162 @@ def text_lines_from_file(filename: Union[str, Path], strip_lines: bool = True) -
...
+    """
+    if os.path.isfile(path):
+        return os.path.getsize(path)
+    total_size = 0
+    for dirpath, dirnames, filenames in os.walk(path):
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            # skip if it is symbolic link
+            if not os.path.islink(fp):
+                total_size += os.path.getsize(fp)
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply allennlp/common/file_utils.py:616]: Hmm good point, should be possible with https://docs.python.org/3.7/library/os.html#os.stat

File: allennlp/common/file_utils.py [hunk 7]
```diff
@@ -7,15 +7,32 @@
...
+    Union,
+    IO,
+    Callable,
+    Set,
+    List,
+    Iterator,
+    Iterable,
+    Dict,
+    NamedTuple,
+)
```
Comment (epwalsh) [inline allennlp/common/file_utils.py:29]: Ew

**Pull request merged**